File: lib/factorix/runtime/wsl.rb
# frozen_string_literal: true

require "json"
require "open3"

module Factorix
  class Runtime
    # WSL (Windows Subsystem for Linux) runtime environment
    #
    # This implementation inherits from Windows and converts Windows paths
    # to WSL paths. It retrieves Windows environment variables via PowerShell
    # in a single batch operation and converts paths using native Ruby code.
    class WSL < Windows
      # Initialize WSL runtime environment
      #
      # @param path [WSLPath] the path provider (for dependency injection)
      def initialize(path: WSLPath.new) = super

      # WSL-specific path provider
      #
      # This class fetches Windows environment variables via PowerShell in one batch operation and converts Windows paths to WSL paths.
      class WSLPath
        # Default WSL mount root for Windows drives
        MOUNT_ROOT = "/mnt"
        private_constant :MOUNT_ROOT

        # PowerShell script to fetch Windows environment variables
        POWERSHELL_SCRIPT = <<~POWERSHELL
          [pscustomobject]@{
            "ProgramFiles(x86)" = ${Env:ProgramFiles(x86)};
            "APPDATA"           = ${Env:APPDATA};
            "LOCALAPPDATA"      = ${Env:LOCALAPPDATA}
          } | ConvertTo-Json -Compress
        POWERSHELL
        private_constant :POWERSHELL_SCRIPT

        POWERSHELL_FALLBACK_PATHS = %w[
          /mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe
          /mnt/c/Windows/system32/WindowsPowerShell/v1.0/powershell.exe
        ].freeze
        private_constant :POWERSHELL_FALLBACK_PATHS

        # Initialize the path provider
        #
        # Sets up the mutex for thread-safe lazy initialization
        def initialize
          @mutex = Mutex.new
        end

        # Get the Program Files (x86) directory path (WSL-converted)
        #
        # @return [Pathname] the Program Files (x86) directory
        def program_files_x86 = @program_files_x86 ||= Pathname(convert_windows_to_wsl(windows_paths["ProgramFiles(x86)"]))

        # Get the AppData directory path (WSL-converted)
        #
        # @return [Pathname] the AppData directory
        def app_data = @app_data ||= Pathname(convert_windows_to_wsl(windows_paths["APPDATA"]))

        # Get the Local AppData directory path (WSL-converted)
        #
        # @return [Pathname] the Local AppData directory
        def local_app_data = @local_app_data ||= Pathname(convert_windows_to_wsl(windows_paths["LOCALAPPDATA"]))

        # Fetch and cache all Windows environment variables
        #
        # @return [Hash] the environment variables as a hash
        private def windows_paths
          return @windows_paths if @windows_paths

          @mutex.synchronize do
            @windows_paths ||= fetch_windows_paths_via_powershell
          end
        end

        # Fetch Windows environment variables via PowerShell
        #
        # @return [Hash] the environment variables as a hash
        # @raise [PlatformError] if PowerShell execution fails
        private def fetch_windows_paths_via_powershell
          ps = find_powershell_exe

          stdout, status = Open3.capture2(ps, "-NoProfile", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-Command", POWERSHELL_SCRIPT)

          raise PlatformError, "PowerShell execution failed: #{status}" unless status.success?

          JSON.parse(stdout.encode("UTF-8", invalid: :replace, undef: :replace))
        end

        # Convert Windows path to WSL path
        #
        # @param windows_path [String] Windows-style path to convert
        # @return [String] equivalent WSL path
        # @raise [PlatformError] if the path format is invalid
        private def convert_windows_to_wsl(windows_path)
          raise PlatformError, "Invalid Windows path: #{windows_path}" unless windows_path =~ %r{\A([A-Za-z]):[\\/]?(.*)\z}

          drive = $1.downcase
          path = $2.tr("\\", "/")
          result = "#{MOUNT_ROOT}/#{drive}/#{path}"
          # Normalize: collapse multiple slashes and remove trailing slash
          result.squeeze("/").delete_suffix("/")
        end

        # Find powershell.exe executable
        #
        # @return [String] path to powershell.exe
        # @raise [PlatformError] if powershell.exe is not found
        private def find_powershell_exe
          return "powershell.exe" if system("which", "powershell.exe", out: File::NULL, err: File::NULL)

          POWERSHELL_FALLBACK_PATHS.find {|path| File.exist?(path) } ||
            raise(PlatformError, "powershell.exe not found in PATH or default locations")
        end
      end
    end
  end
end

File: internal/platform/wsl.go
package platform

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
)

// WSL locates Factorio through the Windows side of the system: it fetches
// the Windows environment variables via powershell.exe in one batch and
// converts the paths to their /mnt/<drive> equivalents. Factorix-side
// directories (cache, config, state) live in the Linux home.
type WSL struct {
	windowsEnvs func() (map[string]string, error)
}

// NewWSL returns a WSL platform. The PowerShell fetch behind windowsPath
// runs at most once, memoized via sync.OnceValues.
func NewWSL() *WSL {
	return &WSL{windowsEnvs: sync.OnceValues(fetchWindowsEnvs)}
}

// The Windows environment variables fetched in one PowerShell invocation.
const wslPowerShellScript = `[pscustomobject]@{
  "ProgramFiles(x86)" = ${Env:ProgramFiles(x86)};
  "APPDATA"           = ${Env:APPDATA};
  "LOCALAPPDATA"      = ${Env:LOCALAPPDATA}
} | ConvertTo-Json -Compress`

var wslPowerShellFallbackPaths = []string{
	"/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
	"/mnt/c/Windows/system32/WindowsPowerShell/v1.0/powershell.exe",
}

func (w *WSL) windowsPath(name string) (string, error) {
	envs, err := w.windowsEnvs()
	if err != nil {
		return "", err
	}
	value, ok := envs[name]
	if !ok || value == "" {
		return "", fmt.Errorf("%w: %s", ErrMissingEnv, name)
	}
	return convertWindowsToWSL(value)
}

func fetchWindowsEnvs() (map[string]string, error) {
	ps, err := findPowerShell()
	if err != nil {
		return nil, err
	}
	out, err := exec.Command(ps, "-NoProfile", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-Command", wslPowerShellScript).Output()
	if err != nil {
		return nil, fmt.Errorf("PowerShell execution failed: %w", err)
	}
	var envs map[string]string
	if err := json.Unmarshal(out, &envs); err != nil {
		return nil, fmt.Errorf("cannot parse PowerShell output: %w", err)
	}
	return envs, nil
}

func findPowerShell() (string, error) {
	if path, err := exec.LookPath("powershell.exe"); err == nil {
		return path, nil
	}
	for _, path := range wslPowerShellFallbackPaths {
		if _, err := os.Stat(path); err == nil {
			return path, nil
		}
	}
	return "", fmt.Errorf("powershell.exe not found in PATH or default locations")
}

var windowsPathRE = regexp.MustCompile(`\A([A-Za-z]):[\\/]?(.*)\z`)

// convertWindowsToWSL maps "C:\Users\x" to "/mnt/c/Users/x".
func convertWindowsToWSL(windowsPath string) (string, error) {
	m := windowsPathRE.FindStringSubmatch(windowsPath)
	if m == nil {
		return "", fmt.Errorf("invalid Windows path: %q", windowsPath)
	}
	drive := strings.ToLower(m[1])
	rest := strings.ReplaceAll(m[2], `\`, "/")
	result := "/mnt/" + drive + "/" + rest
	// Collapse duplicate slashes and drop a trailing one.
	for strings.Contains(result, "//") {
		result = strings.ReplaceAll(result, "//", "/")
	}
	return strings.TrimSuffix(result, "/"), nil
}

func (w *WSL) GameExecutablePath() (string, error) {
	root, err := w.windowsPath("ProgramFiles(x86)")
	if err != nil {
		return "", err
	}
	return steamFactorioPath(root, "bin", "x64", "factorio.exe"), nil
}

func (w *WSL) GameUserDir() (string, error) {
	appData, err := w.windowsPath("APPDATA")
	if err != nil {
		return "", err
	}
	return filepath.Join(appData, "Factorio"), nil
}

func (w *WSL) GameDataDir() (string, error) {
	root, err := w.windowsPath("ProgramFiles(x86)")
	if err != nil {
		return "", err
	}
	return steamFactorioPath(root, "data"), nil
}

func (w *WSL) DefaultCacheHomeDir() (string, error) {
	return w.windowsPath("LOCALAPPDATA")
}

func (w *WSL) DefaultConfigHomeDir() (string, error) {
	return w.windowsPath("APPDATA")
}

func (w *WSL) DefaultDataHomeDir() (string, error) {
	return w.windowsPath("LOCALAPPDATA")
}

func (*WSL) DefaultStateHomeDir() (string, error) {
	return homePath(".local", "state")
}

func (w *WSL) DefaultFactorixLogPath() (string, error) {
	stateHome, err := w.DefaultStateHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(stateHome, "factorix", "factorix.log"), nil
}
